spec/score_card_spec.js

```diff
@@ -0,0 +1,169 @@
+    scorecard.recordRoll(4)
+    expect(scorecard.frame.points).toEqual(9)
+  });
+  xit("records the roll score in roll", function(){
+    scorecard.recordRoll(5)
+    scorecard.recordRoll(4)
+    expect(scorecard.frame.roll2.points).toEqual(4)
+  });
+});
+
+describe("spare bonus", function(){
+  it("records first roll points in previous frame in spare bonus", function(){
+    scorecard.recordRoll(5)
+    scorecard.recordRoll(5)
+    scorecard.recordRoll(4)
+    scorecard.recordRoll(3)
+    expect(scorecard.bonusFrame.points).toEqual(14)
+  });
+});
+
+describe("strike bonus", function(){
+  it("records all roll points in previous frame in strike bonus", function(){
+    scorecard.recordRoll(10)
+    scorecard.recordRoll(4)
+    scorecard.recordRoll(2)
+    expect(scorecard.bonusFrame.points).toEqual(16)
+  });
+});
+
+describe("no bonus", function(){
+  it("doesn't return a bonus score with no bonus mode", function(){
+    scorecard.recordRoll(5)
+    scorecard.recordRoll(3)
+    scorecard.recordRoll(4)
+    expect(scorecard.bonusFrame.points).toEqual(8)
+  });
+});
+
+describe("cheat guard", function(){
+  it("doesn't allow a frame score over 10", function(){
+    scorecard.recordRoll(5)
+    scorecard.recordRoll(6)
+    expect(scorecard.frame.points).toEqual(5)
+  });
+});
+
+describe("total", function(){
+  it("total includes bonus points", function(){
+    scorecard.recordRoll(5)
+    scorecard.recordRoll(5)
+    scorecard.recordRoll(5)
+    expect(scorecard.total).toEqual(20)
+  });
+  it("total includes bonus points 2", function(){
+    scorecard.recordRoll(5)
+    scorecard.recordRoll(5)
+    scorecard.recordRoll(5)
+    scorecard.recordRoll(2)
+    scorecard.recordRoll(5)
+    expect(scorecard.total).toEqual(27)
+  });
+});
+
+describe("final frame", function(){
+  it("allows for only 3 rolls on final frame", function(){
+    scorecard.count = 9
+    scorecard.recordRoll(5)
+    scorecard.recordRoll(5)
+    scorecard.recordRoll(5)
+
+  });
+});
+
+describe("perfect score", function(){
+  it("returns 300 points for a perfect game", function(){
+    scorecard.recordRoll(10)
+    console.log(scorecard.bonusFrame)
+    scorecard.recordRoll(10)
+    console.log(scorecard.bonusFrame)
+    scorecard.recordRoll(10)
+    console.log(scorecard.bonusFrame)
+    scorecard.recordRoll(10)
+    console.log(scorecard.bonusFrame)
+    scorecard.recordRoll(10)
+    console.log(scorecard.bonusFrame)
+    scorecard.recordRoll(10)
+    console.log(scorecard.bonusFrame)
+    scorecard.recordRoll(10)
+    console.log(scorecard.bonusFrame)
+    scorecard.recordRoll(10)
+    console.log(scorecard.bonusFrame)
+    scorecard.recordRoll(10)
+    console.log(scorecard.bonusFrame)
+    scorecard.recordRoll(10)
+    console.log(scorecard.bonusFrame)
+    scorecard.recordRoll(10)
+    console.log(scorecard.bonusFrame)
+    console.log(scorecard.total)
+    console.log(scorecard.frame.points)
+    scorecard.recordRoll(10)
+    console.log(scorecard.bonusFrame)
```

Comment:
Any time we need to call more than a couple of methods for test setup, we should be asking ourselves whether we've effectively separated our concerns. Usually a long test setup is a sign of (a) tightly coupled classes, (b) lack of dependency injection or (c) testing the wrong thing. 

Here, I think there's a mix of all three. Do we need to test that it returns 300 for a perfect game? Or is it enough to test that strike bonus gets applied correctly to the first frame, the 9th frame and the 10th frame (i.e. the three different cases). I would argue that the second option is all that is required, and is in fact better because if the tests fail it's immediately clear where the hole in the logic is. 

But to achieve that test, the way your code is currently set up, you'd need to play 8 frames to set up the 9th frame test and 9 frames to reach the 10th frame test. We should be able to inject state into our score card such that our test is something like:

```
var numberOfFramesInGame = 3;
var frames = [new Frame(10)];
var currentFrame = 1;
var scoredCard = new Scorecard(numberOfFramesInGame, currentFrame, frames);
```

With this test setup, we can immediately test the condition that the penultimate frame handles the strike bonus correctly, and we could do the same for the first and last frames in a game.

(Also, try not to check in console log statements)